model/roles.go

```diff
@@ -373,38 +385,117 @@ func (m *RoleManifest) LookupRole(roleName string) *Role {
-						}
-					}
+		for _, roleJob := range role.RoleJobs {
+			var availableProviders []string
+			for availableName, availableProvider := range roleJob.Job.AvailableProviders {
+				availableProviders = append(availableProviders, availableName)
+				providersByType[availableProvider.Type] = append(providersByType[availableProvider.Type], jobProvidesInfo{
+					jobLinkInfo: jobLinkInfo{
+						Name:     availableProvider.Name,
+						Type:     availableProvider.Type,
+						RoleName: role.Name,
+						JobName:  roleJob.Name,
+					},
+					Properties: availableProvider.Properties,
+				})
+			}
+			for name, provider := range roleJob.ExportedProviders {
+				info, ok := roleJob.Job.AvailableProviders[name]
+				if !ok {
+					errors = append(errors, validation.NotFound(
+						fmt.Sprintf("roles[%s].jobs[%s].provides[%s]", role.Name, roleJob.Name, name),
+						fmt.Sprintf("Provider not found; available providers: %v", availableProviders)))
+					continue
+				}
+				if provider.Alias != "" {
+					name = provider.Alias
 				}
+				providersByName[name] = jobProvidesInfo{
+					jobLinkInfo: jobLinkInfo{
+						Name:     info.Name,
+						Type:     info.Type,
+						RoleName: role.Name,
+						JobName:  roleJob.Name,
+					},
+					Properties: info.Properties,
+				}
+			}
+		}
+	}
 
-				// No valid providers
-				if !consumer.Optional {
+	// Resolve the consumers
+	for _, role := range m.Roles {
+		for _, roleJob := range role.RoleJobs {
+			expectedConsumers := make([]jobConsumesInfo, len(roleJob.Job.DesiredConsumers))
+			copy(expectedConsumers, roleJob.Job.DesiredConsumers)
+			if roleJob.ResolvedConsumers == nil {
+				roleJob.ResolvedConsumers = make(map[string]jobConsumesInfo)
+			}
+			// Deal with any explicitly marked consumers in the role manifest
+			for consumerName, consumerInfo := range roleJob.ResolvedConsumers {
+				consumerAlias := consumerName
+				if consumerInfo.Alias != "" {
+					consumerAlias = consumerInfo.Alias
+				}
+				if consumerAlias == "" {
+					// There was a consumer with an explicitly empty name
+					errors = append(errors, validation.Invalid(
+						fmt.Sprintf(`role[%s].job[%s]`, role.Name, roleJob.Name),
+						"name",
+						fmt.Sprintf("consumer has no name")))
+					continue
+				}
+				provider, ok := providersByName[consumerAlias]
+				if !ok {
 					errors = append(errors, validation.NotFound(
-						fmt.Sprintf(`role[%s].job[%s].consumes[%d]`, role.Name, job.Name, idx),
-						consumer,
-					))
+						fmt.Sprintf(`role[%s].job[%s].consumes[%s]`, role.Name, roleJob.Name, consumerName),
+						fmt.Sprintf(`consumer %s not found`, consumerAlias)))
+					continue
+				}
+				roleJob.ResolvedConsumers[consumerName] = jobConsumesInfo{
+					jobLinkInfo: provider.jobLinkInfo,
+				}
+
+				for i := len(expectedConsumers) - 1; i >= 0; i-- {
+					if expectedConsumers[i].Name == consumerName {
+						expectedConsumers = append(expectedConsumers[:i], expectedConsumers[i+1:]...)
```

Comment:
How can the same `consumerName` appear multiple times?

If it can't, why not add a `break` (and don't walk backwards because it doesn't make any difference then).